File: Hard/72. Edit Distance/EditDistance.java
'''
LeetCode Hard: 72. Edit Distance
https://leetcode.com/problems/edit-distance/

Check EditDistance.cpp for brief solution explanation.
'''

class Solution {
    public int minDistance(String word1, String word2) {
        if(word1.length()<word2.length()){
            String tmp = word1;
            word1 = word2;
            word2 = tmp;
        }
        int N = word1.length(), M = word2.length();
        int[][]dp = new int[2][M+1];
        int pre = 0, cur = 1;
        for(int i = 0;i<=M;i++){
            dp[pre][i] = i;
        }
        for(int i = 1;i<=N;i++){
            dp[cur][0] = i;
            for(int j = 1;j<=M;j++){
                if(word1.charAt(i-1)==word2.charAt(j-1)){
                    dp[cur][j] = dp[pre][j-1];
                    continue;
                }
                dp[cur][j] = 0x3f3f3f3f;
                dp[cur][j] = Math.min(dp[cur][j], dp[pre][j-1]+1);
                dp[cur][j] = Math.min(dp[cur][j], dp[pre][j]+1);
                dp[cur][j] = Math.min(dp[cur][j], dp[cur][j-1]+1);
            }
            pre^=1;
            cur^=1;
        }
        pre^=1;
        cur^=1;
        return dp[cur][M];
    }
}

File: Hard/72. Edit Distance/EditDistance.py
'''
LeetCode Hard: 72. Edit Distance
https://leetcode.com/problems/edit-distance/

Check EditDistance.cpp for brief solution explanation.
'''

class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        if len(word1)<len(word2):
            word1, word2 = word2, word1
        N, M = len(word1), len(word2)
        dp = [[0 for i in range(M+1)] for i in range(2)]
        pre, cur = 0, 1
        for i in range(M+1):
            dp[pre][i] = i
        for i in range(1, N+1):
            dp[cur][0] = i
            for j in range(1, M+1):
                if word1[i-1]==word2[j-1]:
                    dp[cur][j] = dp[pre][j-1]
                    continue
                dp[cur][j] = 0x3f3f3f3f
                dp[cur][j] = min(dp[cur][j], dp[pre][j-1]+1);
                dp[cur][j] = min(dp[cur][j], dp[pre][j]+1);
                dp[cur][j] = min(dp[cur][j], dp[cur][j-1]+1);
            pre, cur = cur, pre
        pre, cur = cur, pre
        return dp[cur][M]